pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,139 +18,283 @@ import (
-	if err := t.countMatchingPods(ctx, topologyGroup); err != nil {
-		return fmt.Errorf("getting matching pods, %w", err)
+func (t *Topology) MaxNonZeroDomain(requirements v1alpha5.Requirements) string {
+	var requirement sets.Set
+	if requirements.Has(t.Key) {
+		requirement = requirements.Get(t.Key)
+	} else {
+		requirement = sets.NewComplementSet()
 	}
-	return nil
+	maxCount := int32(math.MinInt32)
+	var maxDomain string
+	for domain, count := range t.domains {
+		// we only consider domains that match the pod's node selectors
+		if requirement.Has(domain) {
+			if count > maxCount {
+				maxCount = count
+				maxDomain = domain
+			}
+		}
+	}
+	if maxCount == 0 {
+		return ""
+	}
+	return maxDomain
 }
 
-func (t *Topology) computeLabelCapacity(ctx context.Context, constraints *v1alpha5.Constraints, topologyGroup *TopologyGroup) error {
-	topologyGroup.Register(constraints.Requirements.CapacityTypes().UnsortedList()...)
-	if err := t.countMatchingPods(ctx, topologyGroup); err != nil {
-		return fmt.Errorf("getting matching pods, %w", err)
+func (t *Topology) RegisterDomain(zone string) {
+	if _, ok := t.domains[zone]; !ok {
+		t.domains[zone] = 0
 	}
-	return nil
 }
 
-func (t *Topology) countMatchingPods(ctx context.Context, topologyGroup *TopologyGroup) error {
-	pods := &v1.PodList{}
-	if err := t.kubeClient.List(ctx, pods, TopologyListOptions(topologyGroup.Pods[0].Namespace, &topologyGroup.Constraint)); err != nil {
-		return fmt.Errorf("listing pods, %w", err)
+func (t *Topology) UpdateFromCluster(ctx context.Context, kubeClient client.Client) error {
+	podList := &v1.PodList{}
+
+	// collect the pods from all the specified namespaces (don't see a way to query multiple namespaces
+	// simultaneously)
+	var pods []v1.Pod
+	for ns := range t.namespaces {
+		if err := kubeClient.List(ctx, podList, TopologyListOptions(ns, &t.tc)); err != nil {
+			return fmt.Errorf("listing pods, %w", err)
+		}
+		pods = append(pods, podList.Items...)
 	}
-	for i, p := range pods.Items {
-		if IgnoredForTopology(&pods.Items[i]) {
+
+	for i, p := range pods {
+		if IgnoredForTopology(&pods[i]) {
+			continue
+		}
+		if t.Key == v1.LabelHostname {
+			t.RecordUsage(p.Spec.NodeName)
 			continue
 		}
 		node := &v1.Node{}
-		if err := t.kubeClient.Get(ctx, types.NamespacedName{Name: p.Spec.NodeName}, node); err != nil {
+		if err := kubeClient.Get(ctx, types.NamespacedName{Name: p.Spec.NodeName}, node); err != nil {
 			return fmt.Errorf("getting node %s, %w", p.Spec.NodeName, err)
 		}
-		domain, ok := node.Labels[topologyGroup.Constraint.TopologyKey]
+		domain, ok := node.Labels[t.Key]
 		if !ok {
 			continue // Don't include pods if node doesn't contain domain https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/#conventions
 		}
-		topologyGroup.Increment(domain)
+		t.RecordUsage(domain)
 	}
 	return nil
 }
 
+func (t *Topology) EmptyDomain(reqts v1alpha5.Requirements) string {
```

Comment:
Please use fully spelled variables for readability.
```suggestion
func (t *Topology) EmptyDomain(requirements v1alpha5.Requirements) string {
```

Reply:
Happy to change, but it's not idiomatic go IMO.  https://github.com/golang/go/wiki/CodeReviewComments#variable-names